### ISSUE TITLE
In llms/mlx_lm/tuner/trainer.py. The function default_loss is calculating the loss of the whole question and answer pair instead of the loss of answer A 

### ISSUE THREAD

**Issue by alexC-nonsense4k** [ml-explore/mlx-examples#689]
```
def default_loss(model, inputs, targets, lengths):
    logits, _ = model(inputs)
    logits = logits.astype(mx.float32)
    length_mask = mx.arange(inputs.shape[1])[None, :] < lengths[:, None]
    ce = nn.losses.cross_entropy(logits, targets) * length_mask
    ntoks = length_mask.sum()
    ce = ce.sum() / ntoks
    return ce, ntoks
```
For the training of question-answer pairs (QA pairs), this function calculates not only the loss of the Answer part but also the loss of the Question part, indicating that it is a pretraining process rather than a finetuning process for QA pairs. Should a function and process that support finetuning for QA pairs be implemented? Use -100 to mask the part of Question part.

**Comment by awni** [comment]
I think the answer really depends on what works best for your use case. I encourage you to try both versions and see.

If you find the second version (loss for just the answer) works better and would like to add support for it in MLX LM, let us know. We can reopen this issue and work on it!

**Comment by alexC-nonsense4k** [comment]
Hello, I have read some of the source code in the mlx_lm and mlx Python packages, and made some modifications to the code in mlx_lm.
Here are three problems&nbsp;I identified and my thoughts on how to modify them.


1. In mlx_lm.tuner.datasets.py，the `__getitem__` method in the Dataset class only returns a variable of string type. I believe that returning a string type lacks flexibility and scalability, so I adjusted all the Dataset classes and their subclasses to return dictionary types instead. And to calculate the loss only for the answer parts in subsequent steps, I modified the&nbsp;__getitem__&nbsp;method in the&nbsp;CompletionsDataset&nbsp;class to return an additional value. This value is the length of the sequence obtained by encoding only the question part using&nbsp;chat_template&nbsp;with the tokenizer.（done）
2.I made modifications to the&nbsp;iterate_batches&nbsp;function in&nbsp;mlx_lm.tuner.trainer.py&nbsp;by adding a boolean parameter&nbsp;qa. When&nbsp;qa&nbsp;is true, the loss is calculated only for the tokens that are part of the answers.&nbsp;Based on the sequence length of the question part encoded in the&nbsp;__getitem__&nbsp;method of the&nbsp;CompletionsDataset, we can use -100 as a special index for the question part in the&nbsp;target&nbsp;passed to the&nbsp;default_loss&nbsp;method to achieve a masking effect similar to a mask.(done)
3. Following 2, I use -100 as a special index to mask is beacause in pytorch&nbsp;the cross-entropy function supports using -100 as the ignore index. But After reading part of the source code in mlx, I find in mlx&nbsp;the cross-entropy function supports using -100 as the ignore index.&nbsp;&nbsp;To quickly verify and only calculate the loss for the answer part, better results can be achieved in QA training by implementing the -100 masking logic. Currently, I have only made code modifications in&nbsp;default_loss&nbsp;to implement the masking effect of -100.&nbsp;&nbsp;However, I believe that in order to provide better support for future code extensions, the -100 ignore index should be updated into the computation of the cross-entropy function.


After making the above modifications, I trained the model and found that calculating the loss only for the answer part allows the model to achieve better training results on QA tasks.


I have already made some modifications to the existing code, and I am very keen to provide support for mlx_lm. I hope we can reopen this issue and discuss about it！






2513676979
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "ml-explore/mlx-examples"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年4月18日(星期四) 晚上9:07
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [ml-explore/mlx-examples] In llms/mlx_lm/tuner/trainer.py. The function default_loss is calculating the loss of the whole question and answer pair instead of the loss of answer A  (Issue #689)





 
I think the answer really depends on what works best for your use case. I encourage you to try both versions and see.
 
If you find the second version (loss for just the answer) works better and would like to add support for it in MLX LM, let us know. We can reopen this issue and work on it!
